geopandas/io/file.py

```diff
@@ -83,17 +85,46 @@ def _read_file(filename, bbox=None, mask=None, rows=None, **kwargs):
     """
     if _is_url(filename):
         req = _urlopen(filename)
-        path_or_bytes = req.read()
-        reader = fiona.BytesCollection
-    elif isinstance(filename, io.TextIOBase):
-        path_or_bytes = filename.read()
-        reader = fiona.open
+        data = req.read()
+        if get_filetype(data) == "zip":
+            reader = functools.partial(fiona.BytesCollection, data)
+        else:
+            memfile = fiona.io.MemoryFile(data)
+            reader = memfile.open
+    elif pd.api.types.is_file_like(filename):
+        data = filename.read()
+        data = data.encode("utf-8") if isinstance(data, str) else data
+        if get_filetype(data) == "zip":
+            reader = functools.partial(fiona.BytesCollection, data)
+        else:
+            memfile = fiona.io.MemoryFile(data)
+            reader = memfile.open
     else:
-        path_or_bytes = filename
-        reader = fiona.open
+        # Opening a file via URL above automatically detects a zipped file.
+        # In order to match that behavior, attempt to add a zip scheme if missing.
+        # Casting to a pathlib.Path allows the same check to work for plain strings
+        # and pathlib.Path instances.
+        if pathlib.Path(filename).suffix == ".zip":
+            parsed = fiona.parse_path(str(filename))
+            if isinstance(parsed, fiona.path.ParsedPath):
+                # If fiona returns a parsed path, we can safely look at the scheme
+                # and add zip if necessary.
+                if not parsed.scheme:
+                    parsed.scheme = "zip"
+                elif "zip" not in parsed.scheme.split("+"):
+                    parsed.scheme = "zip+" + parsed.scheme
+                filename = parsed.name
+            elif isinstance(parsed, fiona.path.UnparsedPath) and not str(
+                filename
+            ).startswith("/vsi"):
+                # If fiona is unable to parse the path, it might have a Windows drive
+                # scheme. Try adding zip:// to the front. If the path starts with "/vsi"
+                # it is a legacy GDAL path type, so let it pass unmodified.
+                filename = "zip://" + parsed.name
```

Comment:
This is the bit that tries to match zip file handling of local data with that of remote data. The latter automatically infers a zipped file.

I don't love this code block -- it feels like it has too many checks and branches, but there are unfortunately a lot of cases to try to cover. In particular, we want to formulate correct URIs for
* Posix-style paths
* pathlib paths
* windows-style paths
* GDAL `/vsi` style paths (maybe?)

 The good news is that this logic is more-or-less independent of the file-like-object logic, and could be split into a different discussion if necessary.